File: apps/web/src/core/scoring.js
// Bewertung einer CardVote-Session. Spiegel von apps/api/app/scoring.py —
// gleiche Regeln, weil die Auswertungsseite die Gewichte live beim Tippen neu
// rechnet, PDF und Notenbuch-Brücke aber am Server entstehen. Wer hier etwas
// ändert, ändert es dort mit.
//
// E/G: alle sehen dieselben Fragen. Für ein Kind im G-Kurs zählen nur G-Fragen
// als 100 %; richtige E-Fragen geben Bonus — erst ab zwei richtigen, höchstens
// eine Notenstufe. Falsche E-Antworten zehren nur den Bonus auf, nie die Basis.
// Für ein Kind im E-Kurs zählt alles regulär.
//
// Minuspunkte: falsche Antwort kostet ihr Gewicht, nie unter 0. Karte unten
// lassen = keine Antwort = 0 Punkte, kein Abzug.
import { DEFAULT_SCALE } from "./grades.js";

// Prozentpunkte bis zur nächstbesseren Notenstufe — der Deckel des Bonus.
export function naechsteStufe(pct, scale) {
  const s = scale || DEFAULT_SCALE;
  const hoeher = Object.values(s).map(Number).filter((v) => v > pct).sort((a, b) => a - b);
  return hoeher.length ? Math.max(0, hoeher[0] - pct) : Math.max(0, 100 - pct);
}

// "anwesend" | "krank". Ohne jede Antwort gilt krank — außer die Lehrkraft hat
// das Kind ausdrücklich auf anwesend gestellt, dann zählt seine 0 mit.
export function statusOf(cardId, hasAnyScan, config) {
  const cfg = config || {};
  const krank = new Set((cfg.krank || []).map(String));
  const anwesend = new Set((cfg.anwesend || []).map(String));
  const key = String(cardId);
  if (anwesend.has(key)) return "anwesend";
  if (krank.has(key)) return "krank";
  return hasAnyScan ? "anwesend" : "krank";
}

// questions: [{ id, correct_answer, niveau, topic_id }] · answers: { [questionId]: "A" | null }
// gefehltTopics: Themen, bei denen dieses Kind gefehlt hat — ihre Fragen zaehlen
// nicht zur Basis, sondern geben Bonus wie E-Fragen (siehe app/scoring.py).
export function bewerte(questions, answers, { niveau = "", niveauAktiv = false, minuspunkte = false, weights = {}, scale = null, gefehltTopics = null } = {}) {
  const s = scale || DEFAULT_SCALE;
  const gewicht = (qid) => Number(weights[String(qid)] ?? weights[qid] ?? 1) || 0;
  const gegeben = (q) => answers[q.id] ?? answers[String(q.id)] ?? null;
  const richtig = (q) => { const a = gegeben(q); return !!(a && q.correct_answer && q.correct_answer.includes(a)); };

  const zaehlend = questions.filter((q) => q.correct_answer);
  const differenziert = !!niveauAktiv && niveau !== "E";
  const fehlt = new Set((gefehltTopics || []).map(Number).filter((x) => Number.isFinite(x)));
  const verpasst = (q) => fehlt.size > 0 && q.topic_id != null && fehlt.has(Number(q.topic_id));
  const istExtra = (q) => verpasst(q) || (differenziert && (q.niveau || "") === "E");
  let basis = zaehlend.filter((q) => !istExtra(q));
  let extra = zaehlend.filter(istExtra);
  // Keine Basis mehr, weil ein verpasstes Thema alles herausgenommen hat: dann
  // haengt der Bonus an nichts, die verpassten Fragen zaehlen regulaer. Der
  // reine E/G-Fall (G-Kind ohne eine einzige G-Frage) bleibt unberuehrt.
  if (!basis.length && extra.some(verpasst)) {
    basis = extra.filter(verpasst);
    extra = extra.filter((q) => !verpasst(q));
  }

  const baseMax = basis.reduce((sum, q) => sum + gewicht(q.id), 0);
  let score = basis.reduce((sum, q) => sum + (richtig(q) ? gewicht(q.id) : 0), 0);
  if (minuspunkte) score -= basis.reduce((sum, q) => sum + (gegeben(q) && !richtig(q) ? gewicht(q.id) : 0), 0);
  score = Math.max(0, score);
  const basePct = baseMax > 0 ? (score / baseMax) * 100 : 0;

  const eRichtig = extra.filter(richtig).length;
  const eFalsch = extra.filter((q) => gegeben(q) && !richtig(q)).length;
  let bonusPct = 0;
  if (extra.length && eRichtig >= 2) {
    const netto = Math.max(0, eRichtig - eFalsch);
    bonusPct = (netto / extra.length) * naechsteStufe(basePct, s);
  }

  return {
    score: Math.round(score * 100) / 100,
    maxScore: Math.round(baseMax * 100) / 100,
    basePct: Math.round(basePct * 10) / 10,
    bonusPct: Math.round(bonusPct * 10) / 10,
    pct: Math.round(Math.min(100, basePct + bonusPct) * 10) / 10,
    eCorrect: eRichtig,
    eWrong: eFalsch,
    eTotal: extra.length,
    gefehltTotal: extra.filter(verpasst).length,
  };
}

File: apps/api/app/scoring.py
"""Bewertung einer CardVote-Session — eine Quelle für die Regeln.

Dieselben Regeln liegen im Frontend in `apps/web/src/core/scoring.js`; wer hier
etwas ändert, ändert es dort mit. Zwei Fassungen, weil die Auswertungsseite die
Gewichte live beim Tippen neu rechnet, die PDF-/Noten-Wege aber am Server
entstehen.

Die Regeln:

* **E/G-Differenzierung** (Quiz-Flag): alle sehen dieselben Fragen. Für ein Kind
  im G-Kurs zählen nur die G-Fragen als 100 %; richtige E-Fragen geben Bonus
  obendrauf. Der Bonus greift erst ab zwei richtigen E-Antworten und hebt
  höchstens um eine Notenstufe. Falsche E-Antworten zehren nur den Bonus auf
  (bis 0), nie die Basispunkte. Für ein Kind im E-Kurs zählen alle Fragen
  regulär.
* **Minuspunkte** (Quiz-Flag): eine falsche Antwort kostet ihr Gewicht, die
  Punktzahl fällt nie unter 0. Wer die Karte unten lässt, antwortet nicht:
  0 Punkte, kein Abzug.
* **Bei dem Thema gefehlt** (je Kind und Session gesetzt): Fragen zu einem
  Thema, das das Kind verpasst hat, zaehlen NICHT zur Basis — es kann nichts
  wissen, was es nie hatte. Richtige Antworten darauf geben Bonus wie eine
  E-Frage, und Minuspunkte greifen dort nie. Beides landet in einem Topf: der
  Bonus ist zusammen auf hoechstens eine Notenstufe gedeckelt, sonst waere
  „hat gefehlt" ein Weg zu einer besseren Note als bei voller Anwesenheit.
  Sind ALLE Fragen betroffen, gaebe es keine Basis mehr, an der ein Bonus
  haengen koennte — dann zaehlen sie regulaer (eine Wertung aus lauter Bonus
  waere keine).
* **Keine Abgabe**: wer gar nichts abgegeben hat, gilt als krank und bleibt aus
  der Wertung — die Lehrkraft kann ihn auf „anwesend" stellen, dann zählt seine
  0 überall mit (siehe status_of).
"""
import math
from typing import Optional

DEFAULT_SCALE = {1: 87, 2: 73, 3: 59, 4: 45, 5: 20, 6: 0}


def kaufmaennisch(x: float, stellen: int) -> float:
    """Kaufmaennisch runden (die halbe Stelle geht nach oben) — wie Math.round
    im Frontend. Pythons round() rundet die halbe Stelle zur geraden Zahl
    ("Bankers Rounding"): 66,25 % wuerde hier 66,2 und dort 66,3 ergeben, und
    dieselbe Arbeit stuende im PDF anders als auf dem Bildschirm."""
    f = 10 ** stellen
    return math.floor(x * f + 0.5) / f


def naechste_stufe(pct: float, scale: dict) -> float:
    """Prozentpunkte bis zur nächstbesseren Notenstufe — der Deckel des Bonus."""
    s = {int(k): v for k, v in (scale or DEFAULT_SCALE).items()}
    grenzen = sorted((v for v in s.values() if v > pct))
    if not grenzen:
        return max(0.0, 100.0 - pct)   # schon in der besten Stufe: bis 100 %
    return max(0.0, grenzen[0] - pct)


def note_aus_pct(pct: float, scale: Optional[dict] = None) -> float:
    """Prozent -> Note, identisch zur Frontend-Skala (core/grades.js).

    Stand vorher in noten.py und wurde von dort aus benutzt; seit der Themenstand
    dieselbe Umrechnung braucht, gehoert sie hierher — zwei Fassungen derselben
    Notenrechnung liefen unweigerlich auseinander, und dann stuende dieselbe
    Leistung je nach Weg als andere Note da.

    Kaufmaennisch gerundet wie im Frontend (Math.round): Pythons round() rundet
    die halbe Stelle zur geraden Zahl, 83,5 % ergaebe 2,2 statt 2,3.
    """
    try:
        s = {int(k): float(v) for k, v in (scale or {}).items()}
        s = {g: s[g] for g in (1, 2, 3, 4, 5, 6)}   # vollstaendig+lesbar? sonst Default
    except (ValueError, TypeError, KeyError):
        s = {int(k): float(v) for k, v in DEFAULT_SCALE.items()}
    ranges = [(1, s[1], 100), (2, s[2], s[1]), (3, s[3], s[2]), (4, s[4], s[3]), (5, s[5], s[4])]
    for grade, lower, upper in ranges:
        if pct >= lower:
            span = upper - lower
            if span <= 0:
                return float(grade)
            return kaufmaennisch(grade + (upper - pct) / span, 1)
    return 6.0


def status_of(card_id: int, has_any_scan: bool, config: Optional[dict]) -> str:
    """"anwesend" oder "krank". Ohne jede Antwort gilt krank — es sei denn, die
    Lehrkraft hat das Kind ausdrücklich auf anwesend gestellt (dann zählt die 0)."""
    cfg = config or {}
    krank = {str(x) for x in (cfg.get("krank") or [])}
    anwesend = {str(x) for x in (cfg.get("anwesend") or [])}
    key = str(card_id)
    if key in anwesend:
        return "anwesend"
    if key in krank:
        return "krank"
    return "anwesend" if has_any_scan else "krank"


def gefehlt_von(card_id, config: Optional[dict]):
    """Themen, bei denen dieses Kind gefehlt hat (aus eval_config).

    `eval_config["gefehlt"]` ist {card_id: [topic_id, ...]} — dieselbe Ablage
    und derselbe Schluessel wie bei krank/anwesend, damit die Auswertung eine
    Stelle behaelt, an der die Ausnahmen eines Kindes stehen.
    """
    roh = ((config or {}).get("gefehlt") or {})
    if not isinstance(roh, dict):
        return []
    werte = roh.get(str(card_id), roh.get(card_id)) or []
    if not isinstance(werte, (list, tuple, set)):
        return []
    out = []
    for x in werte:
        try:
            out.append(int(x))
        except (TypeError, ValueError):
            continue
    return out


def bewerte(questions, answers, *, niveau: str = "", niveau_aktiv: bool = False,
            minuspunkte: bool = False, weights: Optional[dict] = None,
            scale: Optional[dict] = None, gefehlt_topics=None) -> dict:
    """Punkte und Prozent für ein Kind.

    questions: [{"id", "correct_answer", "niveau", "topic_id"}] — "niveau" ist "E" oder "".
    answers:   {question_id: "A"|None}
    niveau:    Kursniveau des Kindes ("E" | "G" | "")
    gefehlt_topics: Themen, bei denen dieses Kind gefehlt hat (topic_ids)
    """
    w = weights or {}
    scale = {int(k): v for k, v in (scale or DEFAULT_SCALE).items()}

    def gewicht(qid) -> float:
        """Gewicht einer Frage. Fehlt es (oder ist es null), gilt 1 — wie im
        Frontend. Unlesbares (Tippfehler in der Konfiguration) gilt als 0 und
        darf die Wertung nicht mit einem Fehler abbrechen."""
        v = w.get(str(qid), w.get(qid))
        if v is None:
            v = 1
        try:
            v = float(v)
        except (TypeError, ValueError):
            return 0.0
        return 0.0 if math.isnan(v) else v

    def beantwortet(q):
        return answers.get(q["id"]) or answers.get(str(q["id"]))

    def richtig(q):
        a = beantwortet(q)
        c = q.get("correct_answer")
        return bool(a and c and a in c)

    zaehlend = [q for q in questions if q.get("correct_answer")]
    # Ohne E/G-Flag oder für ein Kind im E-Kurs zählt alles regulär.
    differenziert = bool(niveau_aktiv) and niveau != "E"
    fehlt = {int(x) for x in (gefehlt_topics or []) if str(x).lstrip("-").isdigit()}

    def verpasst(q):
        tid = q.get("topic_id")
        return bool(fehlt) and tid is not None and int(tid) in fehlt

    def ist_extra(q):
        return verpasst(q) or (differenziert and (q.get("niveau") or "") == "E")

    basis = [q for q in zaehlend if not ist_extra(q)]
    extra = [q for q in zaehlend if ist_extra(q)]
    # Keine Basis mehr, weil ein verpasstes Thema alles herausgenommen hat: dann
    # haengt der Bonus an nichts. Die verpassten Fragen zaehlen dann regulaer.
    # (Der reine E/G-Fall — ein G-Kind ohne eine einzige G-Frage — bleibt davon
    # unberuehrt: dort ist „keine Basis" seit jeher gewollt.)
    if not basis and any(verpasst(q) for q in extra):
        basis = [q for q in extra if verpasst(q)]
        extra = [q for q in extra if not verpasst(q)]

    base_max = sum(gewicht(q["id"]) for q in basis)
    score = sum(gewicht(q["id"]) for q in basis if richtig(q))
    if minuspunkte:
        score -= sum(gewicht(q["id"]) for q in basis if beantwortet(q) and not richtig(q))
    score = max(0.0, score)
    base_pct = (score / base_max * 100) if base_max > 0 else 0.0

    e_richtig = sum(1 for q in extra if richtig(q))
    e_falsch = sum(1 for q in extra if beantwortet(q) and not richtig(q))
    bonus_pct = 0.0
    if extra and e_richtig >= 2:
        netto = max(0, e_richtig - e_falsch)
        anteil = netto / len(extra)
        bonus_pct = anteil * naechste_stufe(base_pct, scale)

    return {
        "score": kaufmaennisch(score, 2),
        "max_score": kaufmaennisch(base_max, 2),
        "base_pct": kaufmaennisch(base_pct, 1),
        "bonus_pct": kaufmaennisch(bonus_pct, 1),
        "pct": kaufmaennisch(min(100.0, base_pct + bonus_pct), 1),
        "e_correct": e_richtig,
        "e_wrong": e_falsch,
        "e_total": len(extra),
        # Wie viele der Bonus-Fragen aus einem verpassten Thema stammen — die
        # Auswertung sagt sonst „E-Bonus", wo gar keine E-Frage im Spiel war.
        "gefehlt_total": sum(1 for q in extra if verpasst(q)),
    }
